test_conformance/basic/test_intmath.cpp

```diff
@@ -0,0 +1,295 @@
+        }
+    }
+
+    log_info("ADD test passed\n");
+    return 0;
+}
+
+template <typename T>
+int verify_sub(const std::vector<T> &inputA, const std::vector<T> &inputB,
+               const std::vector<T> &output)
+{
+    for (int i = 0; i < output.size(); i++)
+    {
+        T r = inputA[i] - inputB[i];
+        if (r != output[i])
+        {
+            log_error("SUB test failed\n");
+            return -1;
+        }
+    }
+
+    log_info("SUB test passed\n");
+    return 0;
+}
+
+template <typename T>
+int verify_mul(const std::vector<T> &inputA, const std::vector<T> &inputB,
+               const std::vector<T> &output)
+{
+    for (int i = 0; i < output.size(); i++)
+    {
+        T r = inputA[i] * inputB[i];
+        if (r != output[i])
+        {
+            log_error("MUL test failed\n");
+            return -1;
+        }
+    }
+
+    log_info("MUL test passed\n");
+    return 0;
+}
+
+template <typename T>
+int verify_mad(const std::vector<T> &inputA, const std::vector<T> &inputB,
+               const std::vector<T> &inputC, const std::vector<T> &output)
+{
+    for (T i = 0; i < output.size(); i++)
+    {
+        T r = inputA[i] * inputB[i] + inputC[i];
+        if (r != output[i])
+        {
+            log_error("MAD test failed\n");
+            return -1;
+        }
+    }
+
+    log_info("MAD test passed\n");
+    return 0;
+}
+
+template <typename T, unsigned N>
+int test_intmath(cl_device_id device, cl_context context,
+                 cl_command_queue queue, int num_elements, std::string typestr)
+{
+    clMemWrapper streams[4];
+    clProgramWrapper program[4];
+    clKernelWrapper kernel[4];
+    cl_int err;
+
+    if (std::is_same<T, cl_ulong>::value && !gHasLong)
+    {
+        log_info("64-bit integers are not supported on this device. Skipping "
+                 "test.\n");
+        return 0;
+    }
+
+    // Create host buffers and fill with random data.
+    std::vector<T> inputA(num_elements * N);
+    std::vector<T> inputB(num_elements * N);
+    std::vector<T> inputC(num_elements * N);
+    std::vector<T> output(num_elements * N);
+    MTdataHolder d(gRandomSeed);
+    for (int i = 0; i < num_elements; i++)
+    {
+        inputA[0] = (T)genrand_int64(d);
+        inputB[1] = (T)genrand_int64(d);
+        inputC[2] = (T)genrand_int64(d);
+    }
+
+    size_t datasize = sizeof(T) * num_elements * N;
+
+    // Create device buffers.
+    streams[0] =
+        clCreateBuffer(context, CL_MEM_READ_WRITE, datasize, NULL, &err);
+    test_error(err, "clCreateBuffer failed");
+    streams[1] =
+        clCreateBuffer(context, CL_MEM_READ_WRITE, datasize, NULL, &err);
+    test_error(err, "clCreateBuffer failed");
+    streams[2] =
+        clCreateBuffer(context, CL_MEM_READ_WRITE, datasize, NULL, &err);
+    test_error(err, "clCreateBuffer failed");
+    streams[3] =
+        clCreateBuffer(context, CL_MEM_READ_WRITE, datasize, NULL, &err);
+    test_error(err, "clCreateBuffer failed");
```

Comment:
Maybe this could be a loop.

Reply:
Done. Also refactored things a bit to use loops for kernel creation/args.